### PR TITLE
feat(web): longpress restoration, enhancements 🐵

### LINKED ISSUE #1025: Feature Request
On non-Keyman iOS, a longpress pulls up the list, and moving horizontally (not necessarily vertically) moves the cursor above your finger. This is Press+Right+Release. The finger moves along at the original key latitude and doesn't need to cover the choices.  
On Keyman, even on a longpress with one alternate option, the motion is always Press+Right+Up+Release, and your finger must cover the target choice to select it. This is one step more complex. If only one longpress is shown, the process would be Press+Up+Release. 
This is a feature request to extend the touch box on Android/iOS downward below the basekey so that the finger doesn't have to move up to select the popup option, only left/right. I actually achieved this through CSS for KMW, before finding dragons.

image

### LINKED ISSUE #1886: chore(web): UX - Make subkeys easier to press by making touch targets much, MUCH bigger (Fitts's Law)
**Is your feature request related to a problem? Please describe.**

I like being able to access lesser-used characters by pressing and holding a key, and selecting the character I want via a _subkey_. However, selecting those small sub keys is rather small.

Sometimes I select try to select the leftmost subkey and overshoot:

Overshooting the left edge of the subkey array

Sometimes I select try to select the rightmost subkey and overshoot:

Overshooting the right edge of the subkey array

Sometimes I select slightly above the short little subkey array:

Overshooting the top-edge of the subkey array

**Describe the solution you'd like**
What I think would be nice, is if the touch target of the leftmost key extended infinitely to the left; likewise, the touch target of the rightmost key should extended infinitely to the right; and all keys' touch targets should extend to the very top of the keyboard. To cancel the press-and-hold of the menu, you would let go on or below the original key. **Bigger targets are easier to press**. **Infinite targets are impossible not to press**.

**Describe alternatives you've considered**

Instead of extending the touch key infinitely, it may extend dramatically to the sides of the visible key; at a certain point selecting far enough from the key, it should cancel.

**Product context**
 - OS: Android (iOS would be affected too).

**Additional context**

This matches the behaviour in SwiftKey. Gboard extends the touch targets infinitely to the right, extends the targets a bit upwards, and a bit to the left; however, after a certain point to the left, you are no longer selecting subkeys in Gboard.

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #1025.
Fixes #1886.

This PR reworks the old longpress handling methods and classes and connects them to the new gesture engine; they're now operable at this stage of integration.  While I was at it, I took the opportunity to vet some related features baked into the engine; subkeys are now selectable by a "nearest key" heuristic, rather than requiring the touchpoint to be directly inside them.

From the referenced issues:

> image

This behavior is now 100% supported in the exact manner suggested by the image.  Once the subkey menu is displayed, horizontal movement will select the key in the same column at the touchpoint in the nearest subkey row.

> Sometimes I select try to select the rightmost subkey and overshoot:
>
> Overshooting the right edge of the subkey array

(The recording above is in a "before" state, obviously.)

Similarly for the left edge and the top edge - for all three cases, the nearest key will remain selected until the touch location wanders "too far" from the subkey menu - approximately 1/4 to 1/3 a key's height.  It's a decent, but not extreme, fudge factor.

[Edit:  this has since been approximately doubled.]

Now...

captured new longpress

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
Notes from today's team meeting demo:

- Moving _very_ far off-edge of longpress should not cancel the menu, just deselect.

Personal note / question:  So, key de-selection without dropping the menu?

- Starting longpress but not moving finger should select default item
   -  Whoops, missed that detail!
- Roaming touch - roaming up is not possible so perhaps we should only roam same-row for consistency
- Flicks disable up-flick for longpress. But could refine to make that only for keyboards that have north/nw/ne flicks

**Comment by mcdurdin** [comment]
> So, key de-selection without dropping the menu?

Yes

**Comment by jahorton** [comment]
Two new updates:

> Moving very far off-edge of longpress should not cancel the menu, just deselect.

Fixed.  Cancellation still occurs if the finger goes too far off the top or sides of _both_ the actual keyboard _and_ the subkey menu.  Aside from that, selection is only maintained near subkeys and is cleared when wandering far while still within the keyboard's bounds.  I also upped the permitted fudge factor (to maintain selection) a bit.

> Roaming touch - roaming up is not possible so perhaps we should only roam same-row for consistency

Alternative idea:  once roaming has begun, just... disable the up-flick shortcut.  After a new base key has been reached, the up-flick shortcut is no longer viable for that keystroke.  It still works, so long as you don't shift the base key.

I opted to go this way partially because I did find it unintuitive when the neighboring key's menu got displayed on quick input attempts - like if I started near the right edge of 'g' and went up-right, sometimes the subkey menu for 'h' would get displayed instead if I actually went right (instead of up-right) when trying to input the shortcut.  Now, the shortcut just won't trigger... rather than triggering for the wrong key.  "Two birds, one stone" - assuming my perspective on this is reasonable.

(Granted, the latter problem is largely due to using a mouse to emulate touch input... I think.)

**Comment by jahorton** [comment]
> Starting longpress but not moving finger should select default item

This has since been fixed as well; turns out, I'd missed a regression in previous PRs that resulted in the `default` subkey field getting overwritten.

File: web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts [hunk 1]
```diff
@@ -112,20 +117,100 @@ export default class SubkeyPopup implements RealizedGesture {
...
+    // Uses the top boundary from `roamBounding` unless the OSK's main element has a more
+    // permissive top boundary.
+    const topPadding = Math.min(baseBounding.top + basePadding - topContainerBounding.top, 0);
+    const sustainBounding = new PaddedZoneSource(topContainer, [topPadding, 0, 0])
+
+    return {
+      targetRoot: this.element,
+      inputStartBounds: vkbd.element,
+      maxRoamingBounds: sustainBounding,
+      itemIdentifier: (coord, target) => {
```
Comment (jahorton) [inline web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts:167]: Essentially, this is [`findNearestKey`]() & its helper, but for subkeys.

https://github.com/keymanapp/keyman/blob/cac6aad30e52e872883e33e7a4052cf85b4d9749/web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts#L64

**Comment by jahorton** [comment]
I know it's been quite a while since the original requests, but I'd still like to ping @eddieantonio and @MattGyverLee on this to see if the changes seem satisfactory.

While the build product won't be directly accessible due to bulid-agent permissions, I _can_ provide y'all a few screen recordings based on the PR's current state.

**Note**: I've set them to be visible via drop-down when clicking the entries below.  Having lots of animated gifs next to each other can be distracting, after all.

<details><summary>Gigantic longpress menu - two rows, second with one less key</summary>

longpress 1

Going beneath the longpress's base key (rather than the subkey menu) will cancel selection, allowing the user to abort the longpress if they release the touch in such a location.
</details>

<details><summary>Fitts law - loose selection range</summary>

longpress 2 - fitts law

</details>

<details><summary>Swift longpresses - working with muscle memory</summary>

longpress 3 - semiflicks

The 's' longpress is there for comparison on the _standard_ longpress wait timing, while the rest trigger and execute far more quickly.  As long as the longpress-shortcut flick is in a generally upward direction - within roughly 67.5 degrees of a purely-upward direction - and the user doesn't shoot extremely far, the shortcut will trigger early.

Accordingly, if you know where the subkey you want is (which I admittedly didn't model terribly well here, noting the massive 'f' overshoot), it becomes possible to quickly use such subkeys by muscle memory.  Of course, this is probably trickier when there are many subkeys: (1) some could lie too far off the side, outside the accepted angle range, and (2) if there are two rows, gotta be careful to land on the side of the correct row.
</details>

File: web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts [hunk 2]
```diff
@@ -112,20 +117,100 @@ export default class SubkeyPopup implements RealizedGesture {
...
+
+    const roamBounding = new PaddedZoneSource(this.element, [
+      // top
+      basePadding,
+      // left, right
+      basePadding,
+      // bottom: ensure the recognition zone includes the row of the base key.
+      // basePadding is already negative, but bottomDistance isn't.
+      bottomDistance > basePadding ? -bottomDistance : basePadding
+    ]);
```
Comment (jahorton) [inline web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts]: This is the section that defines the 'roaming bounds' that enable and disable key selection.  Note the `basePadding` entry - that value in particular is what defines the maximum overshoot we're allowing.

File: web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts [hunk 3]
```diff
@@ -112,20 +117,100 @@ export default class SubkeyPopup implements RealizedGesture {
...
+
+    const roamBounding = new PaddedZoneSource(this.element, [
+      // top
+      basePadding,
+      // left, right
+      basePadding,
+      // bottom: ensure the recognition zone includes the row of the base key.
+      // basePadding is already negative, but bottomDistance isn't.
+      -bottomDistance < basePadding ? -bottomDistance : basePadding
+    ]);
```
Comment (jahorton) [inline web/src/engine/osk/src/input/gestures/browser/subkeyPopup.ts:154]: This is the section that defines the 'roaming bounds' that enable and disable key selection.  Note the `basePadding` entry - that value in particular is what defines the maximum overshoot we're allowing.

Note:  values here are negative for 'external padding' - essentially, CSS 'margin', but positive for 'internal padding' - essentially, CSS 'padding'.

**Comment by jahorton** [comment]
I've noted a few TODO items on #9752 that likely belong as a direct followup to this work... or possibly even as a component of this PR:

- subkey fat-fingering is still disconnected
- there's no 'cancel the subkey menu' / OSK-level gesture cancellation in place as of yet

Given the size of this PR at present... I'll probably do a followup to this one but before #9752 as a remedy.

**Comment by MattGyverLee** [comment]
@jahorton This looks GREAT! Can you show an animated example where the key touched is on the home row or bottom (A or N)?

**Comment by MattGyverLee** [comment]
One small thing, when there is an odd number of columns, one choice is preselected if you don't move your finger. That's great. Most of my keys have only one popup.

When there are an even number of columns, it's ambiguous/fidgety which one will be selected. I don't know how many of us can regularly hit the left or right side of a key.



What do you think about shifting the popup display half a key left or right when there is an even number of columns so that one "default" key is unambiguously selected? I think it could be shifted a half-key left when the key is on the right side of the screen (O) and a half-key right when the key is on the left side of the screen (E).

**Comment by eddieantonio** [comment]
This looks like a fantastic update! Good GIFs. @MattGyverLee's suggestion about popup display placement is worth looking into. Apart from that, I'm afraid it's been far too long since I've seen Keyman code to comment on the actual code. But this will be a nice addition when it lands!

**Comment by jahorton** [comment]
> Why does the O jump up before the menus shows? It's interesting that the original "o" of the base key disappears while the longpress is active. Is that intentional or a styling mismatch of white text on a white background?

If you're talking about the touch-position 'o', that's because of a little shortcut we've long had for longpresses - an upward motion of sufficient distance during the longpress can trigger the subkey menu to appear early.  It does require moving a distance about equal to 1/4 a key's height, mostly in a vertical direction, so you're probably noticing the movement that's triggering the early display.

> When there is an odd number of columns, one choice will always be preselected if you don't move your finger. That's great and repeatable. 

About that - we recently also had this PR go through:  #9496.  It will be possible to specify a 'default' subkey that will be selected before any movement is registered.  While this likely won't "play nice" with the quick-menu up-flick shortcut, if you just hold position, the specified 'default' key will be the pre-selected one.  There's also a bit of "wiggle room" / "noise cancellation" before the default is unselected.  If no default is selected, but the base key has a direct match as a subkey, that one gets picked as default instead.  So...

> Unfortunately, when there are an even number of columns, it's ambiguous/fidgety which one will be selected. I don't know how many of us can regularly hit the left or right side of a key.

... you'd only run into this issue if there's no default subkey specified in KMW 17.0.

That realignment idea is something we could look at, though due to limited time we may have to defer that specific idea to a later version; we still have to get multitaps and flicks working correctly and ensure those have good visual feedback, and those two currently have higher priority.

**Comment by mcdurdin** [comment]
> That realignment idea is something we could look at, though due to limited time we may have to defer that specific idea to a later version; we still have to get multitaps and flicks working correctly and ensure those have good visual feedback, and those two currently have higher priority.

This seems like a very low priority feature for now.

**Review by mcdurdin — APPROVED** [review]

**Pull request merged**